### ISSUE TITLE
Use of flux job attach to implement MPIR debug interface

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-core#2538]
@grondo:

I am looking at `flux job attach` code and I am not sure if I can make use of it to implement MPIR debug interface any longer. It is a good place to hook MPIR into because it watches eventlogs (both main `eventlog` and `guest.exec.eventlog`). However, the continuations are called for each and every state change that has already happened and it is not clear to me if we can program to this behavior for MPIR.

For parallel debuggers, what is needed would be the *current* state, instead.

For example, if the current state of the main `eventlog` is any state prior to `start`, we want to set a watch on `guest.exec.eventlog` and when your upcoming`sync` event is triggered from that log, we would fill `MPIR_Proctable` and call `MPIR_Breakpoint()`. This would support tool launch mode.

On the other hand, if the current state of the main `eventlog` is `start`, this means the job is already running, we fill `MPIR_Proctable` and call `MPIR_Breakpoint()` immediately after doing some santity check on `guest.exec.eventlog`. This would support tool attach mode.

Other states would be invalid for debuggers. Obvious you can't do much when you have no live processes.

Do you think this is a valid assessment? If so, two more questions:

1. Do we have a job event log interface that gets you the current state? This can be faced with a race condition, but the impact of race conditions would be minimal for MPIR.

2. Maybe we still need a new subcommand like `flux job debug` which would perform MPIR based on the "current" state. My guess it will also have to repeat the bulk of what `flux job attach` would do as well if possible? 

My understanding on this problem may still be shallow so please bear with me until I fully understand the design space.

**Comment by dongahn** [comment]
> Maybe we still need a new subcommand like flux job debug which would perform MPIR based on the "current" state. 

Another approach: If we have a job event log interface that tells you if a job event is current, maybe you can specialize the logic in `flux job attach` to do MPIR... Like "if the state is current and relevant to MPIR, perform MPIR" while keeping the rest of `job attach` logic the same...

**Comment by grondo** [comment]
> Another approach: If we have a job event log interface that tells you if a job event is current, maybe you can specialize the logic in flux job attach to do MPIR... Like "if the state is current and relevant to MPIR, perform MPIR" while keeping the rest of job attach logic the same...

Thanks for looking into this. I think I understand the problem you are describing here.

Would it be sufficient to query the job-manager for the current *coarse* job state (`flux_job_state_t`) before watching any eventlogs when running `flux job attach --debug <id>`, and exit with error if `state > FLUX_JOB_RUN` (or jobid does not exist, presumably inactive or invalid). O/w, we know that `state <= FLUX_JOB_RUN` so we can enter the normal `flux job attach` event watching machinery, and trigger MPIR debug once the job hits the start state.

We still have to handle the case where a job exits before MPIR is fully initialized, but I don't see a way around that.

It may also help that the way I had intended to trigger MPIR_proctable generation for a job was to have the tool send a request to the leader job shell, which would in turn build the proctable via RPCs to subsidiary job shells. Instead of placing the serialized MPIR_proctable in the KVS, perhaps the leader shell could just send the proctable in the response to the tool's RPC. This would solve some synchronization issues since the leader shell could reply with error if the job is exiting (or `ENOSYS` if the job fully exits in the meantime).

If it turns out to be more expedient to put the MPIR_proctable into the kvs, then the leader shell would likely send an event to the exec.eventlog after the commit completes, e.g. `mpir-ready` or something.

**Comment by dongahn** [comment]
I was thinking along the line of somehow augmenting `flux_job_event_watch` so that additional info (the state being current or not) can be passed into the continuation...

I will digest your response as well though.

**Comment by grondo** [comment]
> I was thinking along the line of somehow augmenting flux_job_event_watch so that additional info (the state being current or not) can be passed into the continuation...

It may be difficult to define what it means to be a "current" state in the context of `flux_job_event_watch` though. It may require help from the `job-info` module, which incidentally knows the state of all jobs now...

**Comment by grondo** [comment]
BTW, another gotcha with eventlogs is that events feeding in from multiple sources may be recorded in the eventlog slightly out-of-order, especially if some events are being batched and others are committed synchronously. I can't think of a case where this matters for job eventlogs, but something to keep in mind if we try to implement your idea of flagging the "current" event.

**Comment by dongahn** [comment]
> It may be difficult to define what it means to be a "current" state in the context of flux_job_event_watchthough. 

If the event is currently the last entry of the log, it can be deemed as current? (Naively thinking)

**Comment by grondo** [comment]
I think events are delivered one at a time from `job-info` to the eventlog watch client. `job-info` could determine the "last" entry it currently knows about in a chunk of eventlog and flag it (perhaps first sorting on timestamp), but what if there is a eventlog watch update immediately following the currently processed events? The "current" flag is already out of date before it is sent.

I guess this is no different than watching the coarse state from the job-manager, where each state is considered "current."

In that case, I think we should make a change in the`job-info` module, in `src/modules/job-info/watch.c`: `watch_continuation()`. I'd propose instead of constructing eventlog entry responses in place, we push them onto a list sorted by timestamp first, and add a flag to the final list entry.

**Comment by dongahn** [comment]
Sounds good to me. Since devil is likely in the detail, let me see if I can first mock up some code to see if I can add MPIR into job attach if a form of current state detection is provided. This way, I can be more clear what exactly is required. Thanks.

**Comment by dongahn** [comment]
> Would it be sufficient to query the job-manager for the current coarse job state (flux_job_state_t) before watching any eventlogs when running flux job attach --debug <id>, and exit with error if state > FLUX_JOB_RUN (or jobid does not exist, presumably inactive or invalid)? O/w, we know that state <= FLUX_JOB_RUN so we can enter the normal flux job attach event watching machinery, and trigger MPIR debug once the job hits the start state.

This *may* be good enough... (Still need to think through)

What eventlog do you plan to emit "sync"? My guess is the guest exec eventlog? 

What is going to be the sequence of events in that case?

`init->starting->sync->running`?

And if tool launch mode is not asked,  simply `init->starting->running`?

Maybe what comes down to is to extend `attach_exec_event_continuation` with `event=sync` and `event=running` to do MPIR. Of course, we need to make sure MPIR is not performed redundantly at `event=running` if it was already done with `event=sync`...

**Comment by dongahn** [comment]
BTW, one thing that we need to think through is:

1. How do we support "restart"
2. How do we support "kill" cleanly.

**Comment by dongahn** [comment]
> We still have to handle the case where a job exits before MPIR is fully initialized, but I don't see a way around that.

I wouldn't worry too much about this. This is a legitimate race condition. If we can handle this at job attach, we should. Otherwise, the race should be handled gracefully by the tool.

**Comment by grondo** [comment]
> What eventlog do you plan to emit "sync"? My guess is the guest exec eventlog?

Yes, that is the only eventlog to which the shell has write access.

> What is going to be the sequence of events in that case?
> `init->starting->sync->running`?

Currently the "running" event is emitted by the `job-exec` module once all job-shells have been started, so that state would always precede the "sync" state.

Actually it might be more meaningful if the "running" state was emitted by the shell once we start executing all tasks, or at least after the shell-barrier is complete. The current exec.eventlog states are a bit ad-hoc, so this might be a good time to rethink what the most useful and meaningful execution states might be.

I'll open an issue on this question.

**Comment by grondo** [comment]
See #2539

**Comment by dongahn** [comment]
> Actually it might be more meaningful if the "running" state was emitted by the shell once we start executing all tasks, or at least after the shell-barrier is complete. The current exec.eventlog states are a bit ad-hoc, so this might be a good time to rethink what the most useful and meaningful execution states might be.

Yeah, sounds good to me to tighten up the consistency and semantics of events in the guest exec eventlog.

**Comment by dongahn** [comment]
@grondo: Can we have an agreement on the name of the "sync" event so that I can start to mock up the front end code? I can go with "sync" but if you have a preference towards other names, I can go with that too.

**Comment by grondo** [comment]
@dongahn, yeah I don't have an objection to "sync", though maybe we want to namespace it as `debug.sync` or `ptrace.sync` to avoid future conflict? (Though I think the probability of conflict is low)

**Comment by dongahn** [comment]
@grondo: Please see my other comment at https://github.com/flux-framework/flux-core/issues/2539.

If we use either `running` or `sync`, I would prefer to have "running" in the name sync as well.

Is it possible to introduce
`running` and `running.sync` or `running.tool-sync`?

or similar?

**Comment by grondo** [comment]
Remember these are events not states. But if it helps the front end tool I guess we could call it `running.sync` (I'm not sure I understand the need as there is no other meaning of `sync`)

Also, see my update in #2539 as well.

**Comment by grondo** [comment]
@dongahn, here's another idea I came up with while working on #2513 

After #2513, the shell will always emit the `shell.start` event after all shell ranks have exec'd all tasks. What if in the case of `sync` the *context* for this event contains `{"sync":true}` or `{ "tasks-stopped":true }` (in addition to other context keys)?

In the shell, we can allow plugins to append arbitrary content to the context of the shell events to make this happen easily.

**Comment by dongahn** [comment]
@grondo: this sounds good to me. I can program to a single event (`shell.start`) and differentiate the launch vs. attach case, using its context. Seems like this should work.

**Comment by grondo** [comment]
@dongahn, I'm getting close to proposing a PR which adds support to the job shell to collect MPIR_proctable information on-demand and stop tasks in exec (`sync` support) with a `stop-tasks-in-exec` option.

While I'm working on finishing, you can feel free to check out my [shell-mpir](https://github.com/grondo/flux-core/tree/shell-mpir) branch, and give any early comments here.

For testing, I have modified `flux-job attach` to always fetch the `MPIR_proctable` from the leader shell with the following patch
```diff
diff --git a/src/cmd/flux-job.c b/src/cmd/flux-job.c
index 591f9ad..67e7e91 100644
--- a/src/cmd/flux-job.c
+++ b/src/cmd/flux-job.c
@@ -1150,6 +1153,26 @@ void attach_stdin_cb (flux_reactor_t *r, flux_watcher_t *
     }
 }
 
+void print_proctable (flux_future_t *f, void *arg)
+{
+    const char *s;
+    if (flux_rpc_get (f, &s) < 0)
+        log_err ("proctable_get");
+    else
+        log_msg ("proctable=%s\n", s);
+    flux_future_destroy (f);
+}
+
+void get_mpir_proctable (struct attach_ctx *ctx)
+{
+    char topic [128];
+    flux_future_t *f;
+    snprintf (topic, sizeof (topic), "shell-%ju.proctable", (uintmax_t)ctx->id)
+    if (!(f = flux_rpc_pack (ctx->h, topic, ctx->leader_rank, 0, "{}"))
+        || flux_future_then (f, -1., print_proctable, NULL) < 0)
+        log_err_exit ("get_mpir_proctable");
+}
+
 /* Handle an event in the guest.exec eventlog.
  * This is a stream of responses, one response per event, terminated with
  * an ENODATA error response (or another error if something went wrong).
@@ -1180,6 +1203,7 @@ void attach_exec_event_continuation (flux_future_t *f, voi
     if (!strcmp (name, "shell.init")) {
         flux_watcher_t *w;
 
+
         if (json_unpack (context, "{s:i}",
                          "leader-rank", &ctx->leader_rank) < 0)
             log_err_exit ("error decoding shell.init context");
@@ -1187,6 +1211,8 @@ void attach_exec_event_continuation (flux_future_t *f, voi
         /* flux_buffer_read_watcher_create() requires O_NONBLOCK on
          * stdin */
 
+        get_mpir_proctable (ctx);
+
         if ((stdin_flags = fcntl (STDIN_FILENO, F_GETFL)) < 0)
             log_err_exit ("fcntl F_GETFL stdin");
         if (atexit (restore_stdin_flags) != 0)

```

In the case where you want to build a `MPIR_PROCDESC *MPIR_proctable`, there are convenience functions in `src/shell/libpmir.la`:

```c
MPIR_PROCDESC *proctable_get_mpir_proctable (struct proctable *p);
int proctable_get_size (struct proctable *p);
```


For now, the proctable is fetched directly from the leader shell, and is not stored in the KVS. 

The proctable itself has an ad-hoc encoding that is meant to keep the encoded size small, while making it trivial to combine encoded proctables. The result isn't very pretty though:
(64 task job)
```
 proctable={"hosts":[["fluke",[[62,-3],[1,4],[-2,4],[-2,4],[-2,4],[-2,4],[-2,4],[-2,4],-2]]],"execs":[["sleep",[[-1,-63]]]],"ids":[[0,63]],"pids":[[29046,3],[-13020,3],[2817,3],[114,3],[-82,3],[4608,3],[-69,3],[57,3],[-4523,3],[-3074,3],[3050,3],[-3367,3],[3050,3],[-2950,3],[3285,3],[-212,3]]}
```

**Comment by dongahn** [comment]
Thanks @grondo!

**Comment by dongahn** [comment]
> The proctable itself has an ad-hoc encoding that is meant to keep the encoded size small

Great work and extra mileage! 

> The result isn't very pretty though: (64 task job)

The purpose is an optimization so I didn't expect the actual layout would be pretty. What matters for tools is your awesome convenience functions:

```c
MPIR_PROCDESC *proctable_get_mpir_proctable (struct proctable *p);
int proctable_get_size (struct proctable *p);
```

Thanks.

**Comment by grondo** [comment]
Well as I test on fluke I've found some issues, so let me look into that after lunch before you start testing. (single host tests (eg. flux start) do seem to work ok)

**Comment by dongahn** [comment]
Got it.

**Comment by grondo** [comment]
Ok, I think I fixed the bug in the encoding. I pushed the fixups to the branch mentioned above.

**Comment by dongahn** [comment]
@grondo or @chu11:

I have made some good progress adding MPIR support into flux-job attach. But when I tried to implement the following:

> Would it be sufficient to query the job-manager for the current coarse job state (flux_job_state_t) before watching any eventlogs when running flux job attach --debug <id>, and exit with error if state > FLUX_JOB_RUN (or jobid does not exist, presumably inactive or invalid)? O/w, we know that state <= FLUX_JOB_RUN so we can enter the normal flux job attach event watching machinery, and trigger MPIR debug once the job hits the start state.

It looks to me like there isn't a RPC within `job-manager` that allows me to fetch the state for a particular job. I may be wrong, but the handler list I found:

```
static const struct flux_msg_handler_spec htab[] = {
    {
        FLUX_MSGTYPE_REQUEST,
        "job-manager.list",
        list_handle_request,
        0
    },
    {
        FLUX_MSGTYPE_REQUEST,
        "job-manager.raise",
        raise_handle_request,
        FLUX_ROLE_USER
    },
    {
        FLUX_MSGTYPE_REQUEST,
        "job-manager.kill",
        kill_handle_request,
        FLUX_ROLE_USER
    },
    {
        FLUX_MSGTYPE_REQUEST,
        "job-manager.priority",
        priority_handle_request,
        FLUX_ROLE_USER
    },
    FLUX_MSGHANDLER_TABLE_END,
};
```

I can use the `job-manager.list` request to get the list of the jobs and fetch the info on the target job. But that seems to be an overkill. 

Is there another way to get the coarse state of the target job? Maybe we should add a new request handler to fetch the info on a particular job?

**Comment by garlick** [comment]
@dongahn there isn't currently a way to do that either in `job-info` or `job-manager`.   We could look at adding it although but before doing so we should get a bug open with the use case detail.

There are a currently a couple of constraints on the implementation: 
1) `job-manager` only knows about active jobs, so if the job is already inactive, `job-manager` can't tell if it's inactive or never been born.
2) `job-info` knows about all jobs, but there is a tiny lag between when a job is submitted and when `job-info` gets the first state transition event where it could not yet know about a new job.  If we need to, we could fairly easily IMHO close this gap, because elsewhere in the same module, `job-info.lookup` knows how to fetch and parse the job eventlog from the KVS.  It could try that as a last resort if a requested job is unknown, and the requested jobid is numerically greater than any known job id FLUID.

I would just open that bug but wanted to leave room for comment here as to whether it's needed or not first, since I haven't followed the discussion.

**Comment by dongahn** [comment]
Thanks @garlick. I am actually adding `job-manager.info` to make progress. I do believe the imprecise state info would do the job fine because I just need this for state triaging. After this triaging, the precise states will be provided by watching on job shell events.

**Comment by dongahn** [comment]
> I would just open that bug but wanted to leave room for comment here as to whether it's needed or not first, since I haven't followed the discussion.

Let me make some progress and test this functionality using the current implementation. Once I confirm that this is fine, I will open up an issue ticket.

**Comment by dongahn** [comment]
I am going to start testing MPIR support within flux-job on LC systems where we have totalview installed: https://github.com/dongahn/flux-core/tree/mpir_frontend

> I am actually adding job-manager.info to make progress. 

This RPC is at https://github.com/flux-framework/flux-core/commit/30e8caec85ab099558a9cbcf372141729a19ee28 and https://github.com/flux-framework/flux-core/commit/54d4d267f3043eded25d47697a5e47695496d356

**Comment by grondo** [comment]
> Is there another way to get the coarse state of the target job? Maybe we should add a new request handler to fetch the info on a particular job?

> I am actually adding job-manager.info to make progress.

Longer term, I think the job-info service is probably the right place to add this single-job query, with the fix for the small race mentioned by @garlick above.

Perhaps the existing `job-info.list` RPC could be extended to take a `jobid` argument, indicating the jobid at which to start listing jobs. With `jobid` set and `max_entries == 1` then info on just a single jobid would be sent. (Alternately, an array of jobids could be supported)

**Comment by grondo** [comment]
Similar idea to extend `job-info.list` RPC with `jobid` or arrays of `jobids` was already expressed in #2042.

**Comment by dongahn** [comment]
FYI -- I ran an initial set of tests with totalview under my Mac osx/docker environment.

After a minor tweak from the initial implementation to handle nonexistent`{sync: true}` payload of `shell.init` for attach mode, this got me to debug a simple job for both attach and launch modes (and fullset- and subset-ting).

It appears, however, there are some issues I need to resolve to declare full support. I am going to open up an issue for each to track it. More specifically, we will need
- So-called kill support within totalview
- So-called continue-on-exec-events support within totalview
- Ways to cause `flux mini run` to enable `-o stop-tasks-in-exec` automatically when it senses it is under debugger control

**Comment by dongahn** [comment]
Current launch command line:

```console
$ totalview -oldui -verbosity error --args flux mini run -N 2 -n 2 -o stop-tasks-in-exec ./sleep_print 360 5
```

Current attach command lines:

```console
% flux mini submit -N 2 -n 2 ./sleep_print 360 5
16319265112064
% totalview -oldui -verbosity error --args flux job attach 16319265112064
```

**Comment by dongahn** [comment]
I forgot. We will also need bulk debugger server launch support. (This can be future work though)

**Comment by grondo** [comment]
@dongahn, can this be closed since we merged your PR #2654?
